components/script/dom/gamepadhapticactuator.rs

```diff
@@ -0,0 +1,398 @@
+                }
+            }),
+        );
+
+        let event =
+            EmbedderMsg::StopGamepadHapticEffect(self.gamepad_index as usize, effect_stop_sender);
+        self.global().as_window().send_to_embedder(event);
+
+        self.playing_effect_promise.borrow().clone().unwrap()
+    }
+}
+
+impl GamepadHapticActuator {
+    pub fn has_playing_effect_promise(&self) -> bool {
+        self.playing_effect_promise.borrow().is_some()
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#dom-gamepadhapticactuator-playeffect>
+    /// We are in the task queued by the "in-parallel" steps.
+    pub fn handle_haptic_effect_completed(&self) {
+        if self.effect_sequence_id.get() != self.sequence_id.get() {
+            return;
+        }
+        let playing_effect_promise = self.playing_effect_promise.borrow_mut().take();
+        if let Some(promise) = playing_effect_promise {
+            let message = DOMString::from("complete");
+            promise.resolve_native(&message);
+        }
+    }
+
+    pub fn handle_haptic_effect_stopped(&self, stopped_successfully: bool) {
+        if !stopped_successfully {
+            return;
+        }
+
+        let playing_effect_promise = self.playing_effect_promise.borrow_mut().take();
+
+        if let Some(promise) = playing_effect_promise {
+            let trusted_promise = TrustedPromise::new(promise);
+            let sequence_id = self.sequence_id.get();
+            let reset_sequence_id = self.reset_sequence_id.get();
+            let _ = self.global().gamepad_task_source().queue(
+                task!(complete_promise: move || {
+                    if sequence_id != reset_sequence_id {
+                        warn!("Mismatched sequence/reset sequence ids: {} != {}", sequence_id, reset_sequence_id);
+                        return;
+                    }
+                    let promise = trusted_promise.root();
+                    let message = DOMString::from("complete");
+                    promise.resolve_native(&message);
+                }),
+                &self.global(),
+            );
+        }
+    }
+
+    /// <https://www.w3.org/TR/gamepad/#handling-visibility-change>
+    pub fn handle_visibility_change(&self) {
+        if self.playing_effect_promise.borrow().is_none() {
+            return;
+        }
+
+        let trusted_promise = TrustedPromise::new(
+            self.playing_effect_promise
+                .borrow()
+                .clone()
+                .expect("Promise is null!"),
+        );
+
+        let this = Trusted::new(&*self);
+
+        let _ = self.global().gamepad_task_source().queue(
+            task!(stop_playing_effect: move || {
+                let promise = trusted_promise.root();
+                let actuator = this.root();
+                let message = DOMString::from("preempted");
+                promise.resolve_native(&message);
+                *actuator.playing_effect_promise.borrow_mut() = None;
```

Comment:
Here we actually don't want to use `trusted_promise`, but rather take `playing_effect_promise`, and return early if none.